lahja/asyncio/endpoint.py

```diff
@@ -735,22 +617,36 @@ def stop(self) -> None:
             return
 
         # Broadcast to every connected Endpoint that is allowed to receive the event
+        if remotes_for_broadcast is None:
+            await self._send_to_remotes(
+                item,
+                config,
+                *(
+                    remote
+                    for remote in self._connections
+                    if remote.can_send_item(item, config)
+                ),
+            )
+        else:
+            await self._send_to_remotes(item, config, *remotes_for_broadcast)
+
+    async def _send_to_remotes(
+        self,
+        item: BaseEvent,
+        config: Optional[BroadcastConfig],
+        *remotes: RemoteEndpointAPI,
+    ) -> None:
+        if not item.is_bound:
+            raise Exception("TODO: item must be bound")
         compressed_item = self._compress_event(item)
-        disconnected_endpoints = []
-        for name, remote in list(self._full_connections.items()):
-            # list() makes a copy, so changes to _outbount_connections don't cause errors
-            if remote.can_send_item(item, config):
-                try:
-                    await remote.send_message(Broadcast(compressed_item, config))
-                except RemoteDisconnected:
-                    self.logger.debug(f"Remote endpoint {name} no longer exists")
-                    disconnected_endpoints.append(name)
 
-        if disconnected_endpoints:
-            async with self._remote_connections_changed:
-                for name in disconnected_endpoints:
-                    self._full_connections.pop(name, None)
-                self._remote_connections_changed.notify_all()
+        message = Broadcast(compressed_item, config)
+        for remote in remotes:
+            try:
+                await remote.send_message(message)
+            except RemoteDisconnected:
+                self.logger.debug(f"Remote endpoint {remote} no longer connected")
+                remote.stop_nowait()
```

Comment:
I think `stop_nowait()` could be dropped and this call turned into a call to `stop()` (which doesn't need to be async). Right now it's unclear which one you should call in different situations, and `stop_nowait()` knows a *lot* about the caller, it only works because of how `_run_remote_endpoint()` is written.